### ISSUE TITLE
Cargo make is slow when using workspaces

### ISSUE THREAD

**Issue by edmorley** [sagiegurari/cargo-make#584]
### Describe The Bug
I would love to be able to use cargo make with projects making use of workspaces, to run commands per package (ie not use `workspace = false`), however it's currently too slow. 

As such, I have to use `workspace = false` and instead have a workspace-wide task iterate through the packages myself, which means I can't take advantage of things like `CARGO_MAKE_BINARY_EXECUTABLE_NAME` in order to save having to either hardcode the binary name, or look it up manually.

### To Reproduce
1. In an empty directory, create a `Cargo.toml` and `Makefile.toml` with contents shown below.
2. Run: `for n in $(seq 10); do cargo new "packages/example${n}"; done`
3. Run `time cargo make dist`
4. Run `time cargo make dist-workspace-false`

```toml
# Cargo.toml
[workspace]
members = [
  "packages/*",
]
```

```toml
# Makefile.toml
[env]
CARGO_MAKE_EXTEND_WORKSPACE_MAKEFILE = true

[tasks.dist]
script = '''
echo "Performing packaging step for: $CARGO_MAKE_CRATE_NAME"
# ...
'''

[tasks.dist-workspace-false]
workspace = false
script = '''
IFS=","
for crate_path in ${CARGO_MAKE_CRATE_WORKSPACE_MEMBERS}; do
  echo "Performing packaging step for: $crate_path"
  # ...
done
'''
```

### Expected
The wall time of both of the commands is fairly similar (within a second or so).

### Actual
The `workspace = false` task is 15x slower, making it virtually unusable for situations where the task itself only takes milliseconds (eg hardlinking the binary as part of packaging).

```
$ time cargo make dist
...
[cargo-make] INFO - Build Done in 12.31 seconds.

real	0m13.139s
user	0m9.659s
sys	0m2.200s
```

```
$ time cargo make dist-workspace-false
...
[cargo-make] INFO - Build Done in 0.04 seconds.

real	0m0.850s
user	0m0.618s
sys	0m0.142s
```

For context `time bash -c 'echo hello'` takes 10ms on my machine, so the impact of starting 10 bash shells instead of 1 should be minimal.

From reading #44 it sounds like Cargo make uses subprocesses to run the command against each package. However even bearing that in mind, it is still very slow - given that `time cargo make --version` only takes 135ms on my machine, and 10*135ms doesn't come close to the 13 seconds seen above.

It seems like a few things would help with this:
1. Improve the overall startup time of cargo make when run as a subprocess
2. Provide an option to run per-package tasks in-process
3. Provide the ability to run per-package tasks in parallel with each other

Many thanks :-)

### Versions

```
$ cargo --version
cargo 1.54.0 (5ae8d74b3 2021-06-22)

$ rustup show
Default host: x86_64-apple-darwin
rustup home:  /Users/emorley/.rustup

installed targets for active toolchain
--------------------------------------

x86_64-apple-darwin
x86_64-unknown-linux-musl

active toolchain
----------------

stable-x86_64-apple-darwin (default)
rustc 1.54.0 (a178d0322 2021-07-26)
```

**Comment by sagiegurari** [comment]
can you run it with '--time-summary'? it shows breaks downs where time is spent and might help understand it better.
the to parent process still needs to parse all toml files, and crate level processes might need more complext toml parsing, but it doesn't explain x15

**Comment by edmorley** [comment]
Hi! Thank you for the reply :-)

Using `--time-summary` I get:

```
$ time cargo make --time-summary dist-workspace-false
[cargo-make] INFO - cargo make 0.35.0
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: dist-workspace-false
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: legacy-migration
[cargo-make] INFO - Running Task: dist-workspace-false
Performing packaging step for: packages/example1
Performing packaging step for: packages/example10
Performing packaging step for: packages/example2
Performing packaging step for: packages/example3
Performing packaging step for: packages/example4
Performing packaging step for: packages/example5
Performing packaging step for: packages/example6
Performing packaging step for: packages/example7
Performing packaging step for: packages/example8
Performing packaging step for: packages/example9
[cargo-make] INFO - ================Time Summary================
[cargo-make] INFO - dist-workspace-false:   95.00%	   0.04 seconds
[cargo-make] INFO - legacy-migration:       5.00%	   0.00 seconds
[cargo-make] INFO - ============================================
[cargo-make] INFO - Build Done in 0.04 seconds.

real	0m0.810s
user	0m0.592s
sys	0m0.124s
```

```
$ time cargo make --time-summary dist
[cargo-make] INFO - cargo make 0.35.0
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: dist
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: legacy-migration
[cargo-make] INFO - Running Task: workspace
[cargo-make][1] INFO - Project: example1
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: dist
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Running Task: dist
Performing packaging step for: example1
[cargo-make][1] INFO - ================Time Summary================
[cargo-make][1] INFO - dist:   100.00%	   0.03 seconds
[cargo-make][1] INFO - ============================================
[cargo-make][1] INFO - Build Done in 0.03 seconds.
/Users/emorley/src/testcase-cargo-make
[cargo-make][1] INFO - Project: example10
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: dist
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Running Task: dist
Performing packaging step for: example10
[cargo-make][1] INFO - ================Time Summary================
[cargo-make][1] INFO - dist:   100.00%	   0.03 seconds
[cargo-make][1] INFO - ============================================
[cargo-make][1] INFO - Build Done in 0.03 seconds.
/Users/emorley/src/testcase-cargo-make
[cargo-make][1] INFO - Project: example2
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: dist
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Running Task: dist
Performing packaging step for: example2
[cargo-make][1] INFO - ================Time Summary================
[cargo-make][1] INFO - dist:   100.00%	   0.04 seconds
[cargo-make][1] INFO - ============================================
[cargo-make][1] INFO - Build Done in 0.04 seconds.
/Users/emorley/src/testcase-cargo-make
[cargo-make][1] INFO - Project: example3
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: dist
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Running Task: dist
Performing packaging step for: example3
[cargo-make][1] INFO - ================Time Summary================
[cargo-make][1] INFO - dist:   100.00%	   0.03 seconds
[cargo-make][1] INFO - ============================================
[cargo-make][1] INFO - Build Done in 0.03 seconds.
/Users/emorley/src/testcase-cargo-make
[cargo-make][1] INFO - Project: example4
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: dist
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Running Task: dist
Performing packaging step for: example4
[cargo-make][1] INFO - ================Time Summary================
[cargo-make][1] INFO - dist:   100.00%	   0.03 seconds
[cargo-make][1] INFO - ============================================
[cargo-make][1] INFO - Build Done in 0.03 seconds.
/Users/emorley/src/testcase-cargo-make
[cargo-make][1] INFO - Project: example5
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: dist
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Running Task: dist
Performing packaging step for: example5
[cargo-make][1] INFO - ================Time Summary================
[cargo-make][1] INFO - dist:   100.00%	   0.04 seconds
[cargo-make][1] INFO - ============================================
[cargo-make][1] INFO - Build Done in 0.04 seconds.
/Users/emorley/src/testcase-cargo-make
[cargo-make][1] INFO - Project: example6
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: dist
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Running Task: dist
Performing packaging step for: example6
[cargo-make][1] INFO - ================Time Summary================
[cargo-make][1] INFO - dist:   100.00%	   0.03 seconds
[cargo-make][1] INFO - ============================================
[cargo-make][1] INFO - Build Done in 0.03 seconds.
/Users/emorley/src/testcase-cargo-make
[cargo-make][1] INFO - Project: example7
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: dist
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Running Task: dist
Performing packaging step for: example7
[cargo-make][1] INFO - ================Time Summary================
[cargo-make][1] INFO - dist:   100.00%	   0.04 seconds
[cargo-make][1] INFO - ============================================
[cargo-make][1] INFO - Build Done in 0.04 seconds.
/Users/emorley/src/testcase-cargo-make
[cargo-make][1] INFO - Project: example8
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: dist
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Running Task: dist
Performing packaging step for: example8
[cargo-make][1] INFO - ================Time Summary================
[cargo-make][1] INFO - dist:   100.00%	   0.03 seconds
[cargo-make][1] INFO - ============================================
[cargo-make][1] INFO - Build Done in 0.03 seconds.
/Users/emorley/src/testcase-cargo-make
[cargo-make][1] INFO - Project: example9
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: dist
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Running Task: dist
Performing packaging step for: example9
[cargo-make][1] INFO - ================Time Summary================
[cargo-make][1] INFO - dist:   100.00%	   0.03 seconds
[cargo-make][1] INFO - ============================================
[cargo-make][1] INFO - Build Done in 0.03 seconds.
/Users/emorley/src/testcase-cargo-make
[cargo-make] INFO - ================Time Summary================
[cargo-make] INFO - workspace:          99.98%	   11.64 seconds
[cargo-make] INFO - legacy-migration:   0.02%	   0.00 seconds
[cargo-make] INFO - ============================================
[cargo-make] INFO - Build Done in 11.65 seconds.

real	0m12.418s
user	0m9.314s
sys	0m1.913s
```

**Comment by edmorley** [comment]
I generated a profile using [cargo-flamegraph](https://github.com/flamegraph-rs/flamegraph). It looks a bit sparse (perhaps due to flamegraph-rs/flamegraph/issues/113?) but may be better than nothing for now :-)

```
$ mkdir testcase-cargo-make && cd $_
$ # Create Cargo.lock and Makefile.toml per the steps to reproduce
$ for n in $(seq 10); do cargo new "packages/example${n}"; done
$ cd ..
$ git clone https://github.com/sagiegurari/cargo-make && cd cargo-make
$ cargo install flamegraph
$ sudo CARGO_PROFILE_RELEASE_DEBUG=true cargo flamegraph --bin makers -- --cwd ../testcase-cargo-make/ dist
```
(Had to use `makers` since `cargo-make` panics saying not run via Cargo. However the slow performance is the same when invoking via makers so this is fine. Sudo is required for dtrace to work.)

GitHub won't let me embed the SVG in this comment, so I had to convert it to a PNG (breaking the interactivity/tooltips etc):
<img width="1677" alt="flamegraph-screenshot" src="https://user-images.githubusercontent.com/501702/132465675-d7896364-ea36-40aa-93dd-0675d65d71ab.png">

The original SVG is here:
[flamegraph.svg.gz](https://github.com/sagiegurari/cargo-make/files/7126778/flamegraph.svg.gz)

**Comment by sagiegurari** [comment]
wow, this is really cool. never saw it before.
seems from it that the most time is taken from invoking the git command to pull in the git env/project info.
that doesn't make sense to me since this happens for each sub cargo make invocation as well so i would see it in the run_flow as well * the amount of members. but that one is actually smaller.

**Comment by edmorley** [comment]
I agree it's strange. I'm not quite sure how subprocess invocations of the same program are represented in the flamegraph. Perhaps dtrace/cargo-flamegraph is smart enough to aggregate them?

**Comment by sagiegurari** [comment]
@edmorley I think i found the bottlenecks and gave a solution which i hope is good enough.
few things
* the summary of the build time was in correct. it only included the time the tasks took, but not the time it took cargo-make to initialize. i added to the summary 2 more items (makefile parsing and env setup) and the total time includes the entire cargo-make time.
* most of the time is spent on the setup of the env, that means invoking git, rustc and such to get env info.
However, not everyone needs it. for example if you don't care about git info, i'm now allowing you to skip it.
same for rust info and the crate info.
* in workspace, the setup is done for the parent and each member. this is because things can change. for example you might have a git repo in each member. unlikely but possible.
however in case that is not the case, you can skip the relevant things (for example rust and git) in the members and save up a lot of time.

i wrote a section about it:
https://github.com/sagiegurari/cargo-make/tree/0.35.1#usage-performance-tuning

basically these are the flags:
```toml
[config]
# Skip loading of all core tasks which saves up a bit on toml parsing and task creation
skip_core_tasks = true
# Skips loading Git related environment variables
skip_git_env_info = true
# Skips loading rust related environment variables
skip_rust_env_info = true
# Skips loading the current crate related environment variables
skip_crate_env_info = true
```

would love you to try it out and provide some feedback.

**Comment by sagiegurari** [comment]
@edmorley would love your feedback on this one please

**Comment by sagiegurari** [comment]
closing as this is now released

**Comment by edmorley** [comment]
@sagiegurari Sorry for the delayed reply! Thank you for fixing this - has helped speed things up for me :-)

**Comment by Swoorup** [comment]
Even with the configs, this still takes significant amount of time. Any ideas? @sagiegurari 

```toml
[config]
# Skip loading of all core tasks which saves up a bit on toml parsing and task creation
skip_core_tasks = true
# Skips loading Git related environment variables
skip_git_env_info = true
# Skips loading rust related environment variables
skip_rust_env_info = true
# Skips loading the current crate related environment variables
skip_crate_env_info = true

[tasks.D]
script = "echo hello"
```

```shell
❯ time makers D
[cargo-make] INFO - makers 0.36.3
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: D
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: D
hello
[cargo-make] INFO - Build Done in 0.55 seconds.

________________________________________________________
Executed in  575.91 millis    fish           external
   usr time  308.01 millis  122.00 micros  307.89 millis
   sys time  167.40 millis  938.00 micros  166.46 millis

```

**Comment by sagiegurari** [comment]
lots of stuff going on, on the boot phase. for example cli args parsing and even without core tasks there is a basic toml loaded and so on... 
i guess most of the time is spent before the measurements start so a bit hard to tell at this time, but can dig into it a bit more.

i never invested too much in perf since at the end, cargo-make invokes other commands so overhead improvements seemed minimal to the overall experience. for example, running build or test would take up 99% of the time so there are limits on how i can speed things up on my end. 
anyhow, if you want to investigate this more, need to first open a new issue because old ones will get lost and its hard to track. also feel free to also dig into it as well, as that would greatly help.
